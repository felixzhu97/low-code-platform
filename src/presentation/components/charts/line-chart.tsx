"use client";

import React from "react";
import {
  LineChart as RechartsLineChart,
  Line,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
  ResponsiveContainer,
} from "recharts";

interface LineChartProps {
  data?: Array<Record<string, any>>;
  width?: number;
  height?: number;
  xAxisKey?: string;
  dataKeys?: string[];
  colors?: string[];
  title?: string;
  xField?: string;
  seriesField?: string;
  yField?: string;
}

// 默认数据
const defaultData = [
  { name: "1月", visits: 12500, sales: 8400 },
  { name: "2月", visits: 18900, sales: 12600 },
  { name: "3月", visits: 25600, sales: 16800 },
  { name: "4月", visits: 28400, sales: 19200 },
  { name: "5月", visits: 32100, sales: 22500 },
  { name: "6月", visits: 29800, sales: 20100 },
  { name: "7月", visits: 35400, sales: 26300 },
];

export function LineChart({
  data,
  width = 400,
  height = 300,
  xAxisKey = "name",
  title,
}: LineChartProps) {
  const chartData = data && data.length > 0 ? data : defaultData;

  return (
    <div className="w-full h-full">
      {title && <h3 className="text-lg font-semibold mb-2">{title}</h3>}
      <ResponsiveContainer width="100%" height={height}>
        <RechartsLineChart
          data={chartData}
          margin={{
            top: 20,
            right: 30,
            left: 20,
            bottom: 5,
          }}
        >
          <CartesianGrid strokeDasharray="3 3" />
          <XAxis dataKey={xAxisKey} />
          <YAxis />
          <Tooltip />
          <Legend />
          <Line
            type="monotone"
            dataKey="visits"
            stroke="#8884d8"
            name="访问量"
          />
          <Line type="monotone" dataKey="sales" stroke="#82ca9d" name="销售" />
        </RechartsLineChart>
      </ResponsiveContainer>
    </div>
  );
}

export default LineChart;
