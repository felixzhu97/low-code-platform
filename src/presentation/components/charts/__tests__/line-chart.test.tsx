import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { LineChart } from "../line-chart";

// Mock recharts
vi.mock("recharts", () => ({
  LineChart: ({ children, data, ...props }: any) => (
    <div data-testid="recharts-line-chart" data-props={JSON.stringify(props)}>
      {children}
      <div data-testid="chart-data">{JSON.stringify(data)}</div>
    </div>
  ),
  Line: ({ dataKey, stroke, name, type, ...props }: any) => (
    <div
      data-testid="recharts-line"
      data-datakey={dataKey}
      data-stroke={stroke}
      data-name={name}
      data-type={type}
      {...props}
    />
  ),
  XAxis: ({ dataKey, ...props }: any) => (
    <div data-testid="recharts-x-axis" data-datakey={dataKey} {...props} />
  ),
  YAxis: (props: any) => <div data-testid="recharts-y-axis" {...props} />,
  CartesianGrid: (props: any) => (
    <div data-testid="recharts-cartesian-grid" {...props} />
  ),
  Tooltip: (props: any) => <div data-testid="recharts-tooltip" {...props} />,
  Legend: (props: any) => <div data-testid="recharts-legend" {...props} />,
  ResponsiveContainer: ({ children, width, height, ...props }: any) => (
    <div
      data-testid="recharts-responsive-container"
      data-width={width}
      data-height={height}
      {...props}
    >
      {children}
    </div>
  ),
}));

describe("LineChart", () => {
  const sampleData = [
    { name: "1月", visits: 12500, sales: 8400 },
    { name: "2月", visits: 18900, sales: 12600 },
    { name: "3月", visits: 25600, sales: 16800 },
    { name: "4月", visits: 28400, sales: 19200 },
  ];

  describe("Rendering", () => {
    it("renders with default props", () => {
      render(<LineChart />);

      expect(
        screen.getByTestId("recharts-responsive-container")
      ).toBeInTheDocument();
      expect(screen.getByTestId("recharts-line-chart")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-x-axis")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-y-axis")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-cartesian-grid")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-tooltip")).toBeInTheDocument();
      expect(screen.getByTestId("recharts-legend")).toBeInTheDocument();
    });

    it("renders with custom data", () => {
      render(<LineChart data={sampleData} />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement).toBeInTheDocument();
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("12500");
    });

    it("renders with title", () => {
      const title = "Visits and Sales Trend";
      render(<LineChart title={title} />);

      expect(screen.getByText(title)).toBeInTheDocument();
    });

    it("renders with custom dimensions", () => {
      render(<LineChart width={700} height={450} />);

      const container = screen.getByTestId("recharts-responsive-container");
      expect(container).toHaveAttribute("data-width", "100%");
      expect(container).toHaveAttribute("data-height", "450");
    });

    it("renders with custom xAxisKey", () => {
      render(<LineChart data={sampleData} xAxisKey="month" />);

      const xAxis = screen.getByTestId("recharts-x-axis");
      expect(xAxis).toHaveAttribute("data-datakey", "month");
    });

    it("renders visits and sales lines", () => {
      render(<LineChart data={sampleData} />);

      const lines = screen.getAllByTestId("recharts-line");
      expect(lines).toHaveLength(2);

      expect(lines[0]).toHaveAttribute("data-datakey", "visits");
      expect(lines[0]).toHaveAttribute("data-stroke", "#8884d8");
      expect(lines[0]).toHaveAttribute("data-name", "访问量");
      expect(lines[0]).toHaveAttribute("data-type", "monotone");

      expect(lines[1]).toHaveAttribute("data-datakey", "sales");
      expect(lines[1]).toHaveAttribute("data-stroke", "#82ca9d");
      expect(lines[1]).toHaveAttribute("data-name", "销售");
      expect(lines[1]).toHaveAttribute("data-type", "monotone");
    });
  });

  describe("Data Handling", () => {
    it("uses default data when no data provided", () => {
      render(<LineChart />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("7月");
    });

    it("uses default data when empty array provided", () => {
      render(<LineChart data={[]} />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("7月");
    });

    it("uses provided data when valid data given", () => {
      render(<LineChart data={sampleData} />);

      const chartDataElement = screen.getByTestId("chart-data");
      expect(chartDataElement.textContent).toContain("1月");
      expect(chartDataElement.textContent).toContain("4月");
      expect(chartDataElement.textContent).not.toContain("7月");
    });
  });

  describe("Props Configuration", () => {
    it("applies correct chart margins", () => {
      render(<LineChart data={sampleData} />);

      const chartElement = screen.getByTestId("recharts-line-chart");
      const props = JSON.parse(chartElement.getAttribute("data-props") || "{}");

      expect(props.margin).toEqual({
        top: 20,
        right: 30,
        left: 20,
        bottom: 5,
      });
    });

    it("sets correct default dimensions", () => {
      render(<LineChart />);

      const container = screen.getByTestId("recharts-responsive-container");
      expect(container).toHaveAttribute("data-width", "100%");
      expect(container).toHaveAttribute("data-height", "300");
    });

    it("uses default xAxisKey when not provided", () => {
      render(<LineChart data={sampleData} />);

      const xAxis = screen.getByTestId("recharts-x-axis");
      expect(xAxis).toHaveAttribute("data-datakey", "name");
    });
  });

  describe("Line Styling", () => {
    it("applies correct line styles", () => {
      render(<LineChart data={sampleData} />);

      const lines = screen.getAllByTestId("recharts-line");

      // Check line types are monotone
      lines.forEach((line) => {
        expect(line).toHaveAttribute("data-type", "monotone");
      });
    });
  });

  describe("Accessibility", () => {
    it("has proper heading structure when title is provided", () => {
      const title = "Monthly Traffic Report";
      render(<LineChart title={title} />);

      const heading = screen.getByRole("heading", { level: 3 });
      expect(heading).toHaveTextContent(title);
    });

    it("does not render heading when no title provided", () => {
      render(<LineChart />);

      const heading = screen.queryByRole("heading");
      expect(heading).not.toBeInTheDocument();
    });
  });

  describe("Snapshot Tests", () => {
    it("matches snapshot with default props", () => {
      const { container } = render(<LineChart />);
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot with custom data and title", () => {
      const { container } = render(
        <LineChart
          data={sampleData}
          title="Custom Line Chart"
          width={600}
          height={400}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it("matches snapshot with all props", () => {
      const { container } = render(
        <LineChart
          data={sampleData}
          width={800}
          height={500}
          xAxisKey="name"
          title="Complete Line Chart"
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});
