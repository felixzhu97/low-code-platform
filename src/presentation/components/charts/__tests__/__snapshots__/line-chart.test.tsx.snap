// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LineChart > Snapshot Tests > matches snapshot with all props 1`] = `
<div>
  <div
    class="w-full h-full"
  >
    <h3
      class="text-lg font-semibold mb-2"
    >
      Complete Line Chart
    </h3>
    <div
      data-height="500"
      data-testid="recharts-responsive-container"
      data-width="100%"
    >
      <div
        data-props="{"margin":{"top":20,"right":30,"left":20,"bottom":5}}"
        data-testid="recharts-line-chart"
      >
        <div
          data-testid="recharts-cartesian-grid"
          stroke-dasharray="3 3"
        />
        <div
          data-datakey="name"
          data-testid="recharts-x-axis"
        />
        <div
          data-testid="recharts-y-axis"
        />
        <div
          data-testid="recharts-tooltip"
        />
        <div
          data-testid="recharts-legend"
        />
        <div
          data-datakey="visits"
          data-name="访问量"
          data-stroke="#8884d8"
          data-testid="recharts-line"
          data-type="monotone"
        />
        <div
          data-datakey="sales"
          data-name="销售"
          data-stroke="#82ca9d"
          data-testid="recharts-line"
          data-type="monotone"
        />
        <div
          data-testid="chart-data"
        >
          [{"name":"1月","visits":12500,"sales":8400},{"name":"2月","visits":18900,"sales":12600},{"name":"3月","visits":25600,"sales":16800},{"name":"4月","visits":28400,"sales":19200}]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LineChart > Snapshot Tests > matches snapshot with custom data and title 1`] = `
<div>
  <div
    class="w-full h-full"
  >
    <h3
      class="text-lg font-semibold mb-2"
    >
      Custom Line Chart
    </h3>
    <div
      data-height="400"
      data-testid="recharts-responsive-container"
      data-width="100%"
    >
      <div
        data-props="{"margin":{"top":20,"right":30,"left":20,"bottom":5}}"
        data-testid="recharts-line-chart"
      >
        <div
          data-testid="recharts-cartesian-grid"
          stroke-dasharray="3 3"
        />
        <div
          data-datakey="name"
          data-testid="recharts-x-axis"
        />
        <div
          data-testid="recharts-y-axis"
        />
        <div
          data-testid="recharts-tooltip"
        />
        <div
          data-testid="recharts-legend"
        />
        <div
          data-datakey="visits"
          data-name="访问量"
          data-stroke="#8884d8"
          data-testid="recharts-line"
          data-type="monotone"
        />
        <div
          data-datakey="sales"
          data-name="销售"
          data-stroke="#82ca9d"
          data-testid="recharts-line"
          data-type="monotone"
        />
        <div
          data-testid="chart-data"
        >
          [{"name":"1月","visits":12500,"sales":8400},{"name":"2月","visits":18900,"sales":12600},{"name":"3月","visits":25600,"sales":16800},{"name":"4月","visits":28400,"sales":19200}]
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LineChart > Snapshot Tests > matches snapshot with default props 1`] = `
<div>
  <div
    class="w-full h-full"
  >
    <div
      data-height="300"
      data-testid="recharts-responsive-container"
      data-width="100%"
    >
      <div
        data-props="{"margin":{"top":20,"right":30,"left":20,"bottom":5}}"
        data-testid="recharts-line-chart"
      >
        <div
          data-testid="recharts-cartesian-grid"
          stroke-dasharray="3 3"
        />
        <div
          data-datakey="name"
          data-testid="recharts-x-axis"
        />
        <div
          data-testid="recharts-y-axis"
        />
        <div
          data-testid="recharts-tooltip"
        />
        <div
          data-testid="recharts-legend"
        />
        <div
          data-datakey="visits"
          data-name="访问量"
          data-stroke="#8884d8"
          data-testid="recharts-line"
          data-type="monotone"
        />
        <div
          data-datakey="sales"
          data-name="销售"
          data-stroke="#82ca9d"
          data-testid="recharts-line"
          data-type="monotone"
        />
        <div
          data-testid="chart-data"
        >
          [{"name":"1月","visits":12500,"sales":8400},{"name":"2月","visits":18900,"sales":12600},{"name":"3月","visits":25600,"sales":16800},{"name":"4月","visits":28400,"sales":19200},{"name":"5月","visits":32100,"sales":22500},{"name":"6月","visits":29800,"sales":20100},{"name":"7月","visits":35400,"sales":26300}]
        </div>
      </div>
    </div>
  </div>
</div>
`;
